compiler/rustc_lint/src/lifetime_style.rs

```diff
@@ -0,0 +1,454 @@
+use rustc_data_structures::fx::FxIndexMap;
+use rustc_hir::intravisit::{self, Visitor};
+use rustc_hir::{self as hir, LifetimeSource, LifetimeSyntax};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::Span;
+use tracing::instrument;
+
+use crate::{LateContext, LateLintPass, LintContext, lints};
+
+declare_lint! {
+    /// The `mismatched_lifetime_syntaxes` lint detects when an
+    /// elided lifetime uses different syntax between function
+    /// arguments and return values.
+    ///
+    /// The three kinds of syntax are:
+    /// 1. Explicitly bound lifetimes, such as `'static` or `'a`.
+    /// 2. The explicitly anonymous lifetime `'_`.
+    /// 3. Implicit lifetimes, such as `&u8` or `ThisHasALifetimeGeneric`.
+    ///
+    /// As an exception, this lint allows references with implicit or
+    /// explicitly anonymous lifetimes to be paired with paths using
+    /// explicitly anonymous lifetimes.
```

Comment:
This documentation is user-facing, so it should use the lang model from https://github.com/rust-lang/rust/pull/120808#issuecomment-2741166247.